File: python3/_102_BinaryTreeLevelOrderTraversal.py
from typing import List, Optional

from python3.helper import Eg, int_list_to_tree, tester_helper
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from python3.helper_classes import TreeNode


class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root is None:
            return []
        result = []
        level = [root]
        while len(level) > 0:
            level_vals = []
            next_level = []
            for node in level:
                level_vals.append(node.val)
                if node.left is not None:
                    next_level.append(node.left)
                if node.right is not None:
                    next_level.append(node.right)
            result.append(level_vals)
            level = next_level
        return result


def tester():
    examples = [
        Eg(int_list_to_tree([3, 9, 20, None, None, 15, 7]),
           [[3], [9, 20], [15, 7]]),
        Eg(int_list_to_tree([1]), [[1]]),
        Eg(int_list_to_tree([]), [])
    ]
    tester_helper(102, examples, Solution().levelOrder)

File: rust/src/_102_binary_tree_level_order_traversal.rs
//! Solution for https://leetcode.com/problems/binary-tree-level-order-traversal
//! 102. Binary Tree Level Order Traversal

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::collections::VecDeque;
use std::rc::Rc;
impl Solution {
    pub fn level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        let Some(root) = root else {
            return vec![];
        };

        let mut result = vec![];
        let mut queue = VecDeque::new();
        queue.push_back((root, 0));
        while let Some((node, level)) = queue.pop_front() {
            if level >= result.len() {
                // Extend to store new level
                result.push(vec![]);
                debug_assert!(level < result.len());
            }

            result[level].push(node.borrow().val);
            if let Some(left) = node.borrow_mut().left.take() {
                queue.push_back((left, level + 1));
            }
            if let Some(right) = node.borrow_mut().right.take() {
                queue.push_back((right, level + 1));
            }
        }

        result
    }
}

// << ---------------- Code below here is only for local use ---------------- >>

pub struct Solution;
use cargo_leet::TreeNode;

#[cfg(test)]
mod tests {
    use super::*;
    use cargo_leet::TreeRoot;

    use rstest::rstest;

    #[rstest]
    #[case(TreeRoot::from("[3,9,20,null,null,15,7]").into(), vec![vec![3],vec![9,20],vec![15,7]])]
    #[case(TreeRoot::from("[1]").into(), vec![vec![1]])]
    #[case(TreeRoot::from("[]").into(), vec![])]
    fn case(#[case] root: Option<Rc<RefCell<TreeNode>>>, #[case] expected: Vec<Vec<i32>>) {
        let actual = Solution::level_order(root);
        assert_eq!(actual, expected);
    }
}
